pkg/k8s/state_into_spec_annotation.go

```diff
@@ -19,24 +19,54 @@ import (
 	"strings"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-func ValidateOrDefaultStateIntoSpecAnnotation(obj *unstructured.Unstructured) error {
-	_, found := GetAnnotation(StateIntoSpecAnnotation, obj)
-	if !found {
-		SetAnnotation(StateIntoSpecAnnotation, StateMergeIntoSpec, obj)
+// StateIntoSpecValue contains the required 'defaultValue' field and the
+// optional 'userOverride' field.
+type StateIntoSpecValue struct {
+	defaultValue string
+	userOverride *string
+}
+
+func NewStateIntoSpecValue(defaultValue string, userOverride *string) (*StateIntoSpecValue, error) {
+	if !isAcceptedValue(defaultValue, StateIntoSpecAnnotationValues) {
+		return nil, fmt.Errorf("invalid default value '%v' for '%v' annotation, need to be one of {%v}", defaultValue, StateIntoSpecAnnotation, strings.Join(StateIntoSpecAnnotationValues, ", "))
+	}
+	if userOverride != nil && !isAcceptedValue(*userOverride, StateIntoSpecAnnotationValues) {
+		return nil, fmt.Errorf("invalid user override value '%v' for '%v' annotation, need to be one of {%v}", userOverride, StateIntoSpecAnnotation, strings.Join(StateIntoSpecAnnotationValues, ", "))
 	}
-	return validateStateIntoSpecAnnotation(obj, obj.GroupVersionKind())
+	return &StateIntoSpecValue{
+		defaultValue: defaultValue,
+		userOverride: userOverride,
+	}, nil
 }
 
-func EnsureSpecIntoSateAnnotation(obj *Resource) error {
+func (v *StateIntoSpecValue) GetValue() string {
+	if v.userOverride == nil {
+		return v.defaultValue
+	}
+	return *v.userOverride
+}
+
+// ValidateOrDefaultStateIntoSpecAnnotation defaults the annotation to the
+// passed in defaultValue if it is unset and validates the value of the
+// 'state-into-spec' annotation.
+func ValidateOrDefaultStateIntoSpecAnnotation(obj metav1.Object, gvk schema.GroupVersionKind, defaultValue string) error {
 	_, found := GetAnnotation(StateIntoSpecAnnotation, obj)
 	if !found {
+		defaultStateIntoSpecAnnotation(obj, gvk, defaultValue)
+	}
+	return validateStateIntoSpecAnnotation(obj, gvk)
+}
+
+func defaultStateIntoSpecAnnotation(obj metav1.Object, gvk schema.GroupVersionKind, defaultValue string) {
```

Comment:
Nit: We could write a function `GetGVKForKCCType(obj metav1.Object)` or use `runtime.Object`.  controller-runtime has https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client/apiutil#GVKForObject , though that takes a scheme and we could either have a global scheme instance or we could simply hard-code our own logic here.   That saves having to pass the GVK around.